klippy/extras/position_recall.py

```diff
@@ -0,0 +1,52 @@
+# Save and restore position
+#
+# Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+class error(Exception):
+    pass
+
+class PositionRecall:
+    def __init__(self, config):
+        self.slots = {}
+        self.printer = config.get_printer()
+        self.gcode = self.toolhead = None
+    def printer_state(self, state):
+        if state == 'ready':
+            self.gcode = self.printer.lookup_object('gcode')
+            self.toolhead = self.printer.lookup_object('toolhead')
+            self.gcode.register_command("SAVE_CURRENT_POSITION", self.cmd_SAVE_CURRENT_POSITION,
+                                    desc=self.cmd_SAVE_CURRENT_POSITION_help)
+            self.gcode.register_command("RESTORE_POSITION", self.cmd_RESTORE_POSITION,
+                                    desc=self.cmd_RESTORE_POSITION_help)
+    cmd_SAVE_CURRENT_POSITION_help = "Save the current position to a slot"
+    def cmd_SAVE_CURRENT_POSITION(self, params):
+        name = self.gcode.get_str('NAME', params, 'default')
+        self.slots[name] = (self.toolhead.get_position(), self.gcode.speed)
+    cmd_RESTORE_POSITION_help = "Move to the position previously saved in a slot"
+    def cmd_RESTORE_POSITION(self, params):
+        name = self.gcode.get_str('NAME', params, 'default')
+        if name not in self.slots:
+            raise error("Invalid slot '%s' not previously set" % (name,))
```

Comment:
Unless I'm missing something, that should be "raise self.gcode.error(...)".

Reply:
Error is not part of the class and defined at the top level of the module and not part of the PositionRecall class .  Followed your example in another file.  Can change if you would like.